---
layout: post
title: "EOB-002: Toehold switches De-Novo-Designed Regulators of Gene Expression"
description: ""
category: 
- Eyes on biology
tags:
- EOB
- small RNAs
- riboregulators
- prokaryotic
---
{% include JB/setup %}

## Toehold Switches: De-Novo-Designed Regulators of Gene Expression  

> Alexander A. Green,1 Pamela A. Silver,1,2 James J. Collins,1,3 and Peng Yin1,2,  

>    * 1Wyss Institute for Biologically Inspired Engineering, Harvard University, Boston, MA 02115, USA   
>    * 2Department of Systems Biology, Harvard Medical School, Boston, MA 02115, USA   
>    * 3Howard Hughes Medical Institute, Department of Biomedical Engineering and Center of Synthetic Biology, Boston University, Boston, MA 02215, USA   
>    *Correspondence: py@hms.harvard.edu
>    [http://dx.doi.org/10.1016/j.cell.2014.10.002](http://dx.doi.org/10.1016/j.cell.2014.10.002)  

### Highlights

* Toehold switches detect RNAs with arbitrary sequences, including endogenous RNAs   
* A library of 26 orthogonal toehold switches provides <12% component crosstalk   
* Rational RNA sequence design yields switches with >400-fold mean dynamic range   
* Toehold switches can be integrated into the genome and complex genetic circuits   

### Summary
Efforts to construct synthetic networks in living cells have been hindered by the limited number of regulatory components that provide wide dynamic range and low crosstalk. Here, we report a class of de-novo-designed prokaryotic riboregulators called toehold switches that activate gene expression in response to cognate RNAs with arbitrary sequences. Toehold switches provide a high level of orthogonality and can be forward engineered to provide average dynamic range above 400. We show that switches can be integrated into the genome to regulate endogenous genes and use them as sensors that respond to endogenous RNAs. We exploit the orthogonality of toehold switches to regulate 12 genes independently and to construct a genetic circuit that evaluates 4-input AND logic. Toehold switches, with their wide dynamic range, orthogonality, and programmability, represent a versatile and powerful platform for regulation of translation, offering diverse applications in molecular biology, synthetic biology, and biotechnology.

<figure>
    <img src="{{ base.url }}/assets/images/toehold_switches_fig1-1.png" alt="Toehold switches" width=100%>
    <img src="{{ base.url }}/assets/images/toehold_switches_fig1-2.png" alt="Toehold switches" width=100%>
    <figurecaption><b>Figure 1. Toehold Switch Design and In Vivo Characterization</b> (A and B) Design schematics of conventional riboregulators (A) and toehold switches (B). Variable sequences are shown in gray, whereas conserved or constrained sequences are represented by different colors. </br>(A) Conventional riboregulator systems repress translation by base pairing directly to the RBS region. RNA-RNA interactions are initiated via a loop-linear or looploop interaction at the YUNR loop in an RNA hairpin. </br>(B) Toehold switches repress translation through base pairs programmed before and after the start codon (AUG), leaving the RBS and start codon regions completely unpaired. RNA-RNA interactions are initiated via linear-linear interaction domains called toeholds. The toehold domain a binds to a complementary a* domain on the trigger RNA. Domains a and b are 12 and 18 nts, respectively. </br>(C) Flow cytometry GFP fluorescence histograms for toehold switch number 2 compared to E. coli autofluorescence and a positive control. Autofluorescence level measured from induced cells not bearing a GFP-expressing plasmid. </br>(D) GFP mode fluorescence levels measured for switches in their ON and OFF states in comparison to positive control constructs and autofluorescence. Error bars are the SD from at least three biological replicates. </br>(E) ON/OFF GFP fluorescence levels obtained 3 hr after induction for 168 first-generation toehold switches. Inset: ON/OFF GFP fluorescence measured for toehold switches of varying performance levels at different time points following induction. Relative errors for the switch ON/OFF ratios were obtained by adding the relative errors of the switch ON and OFF state fluorescence measurements in quadrature. Relative errors for ON and OFF states are from the SD of at least three biological replicates.</figurecaption>

</figure>
